dali/imgcodec/decoders/nvjpeg_lossless/nvjpeg_lossless.cc

```diff
@@ -33,114 +33,188 @@ NvJpegLosslessDecoderInstance::NvJpegLosslessDecoderInstance(
+    per_thread_resources_.reserve(nthreads);
+    for (int i = per_thread_resources_.size(); i < nthreads; i++)
+      per_thread_resources_.emplace_back(nvjpeg_handle_);
+  }
+  sample_meta_.clear();
+  sample_meta_.resize(nsamples);
+
+  // temporary solution: Just to check when the parsing has finished
+  DecodeResultsPromise parse_promise(nsamples);
+  for (int i = 0; i < nsamples; i++) {
+    int tid = 0;
+    ctx.tp->AddWork(
+        [&, i](int tid) {
+          auto &jpeg_stream = per_thread_resources_[tid].jpeg_stream;
+          auto *sample = in[i];
+          auto &meta = sample_meta_[i];
+          try {
+            CUDA_CALL(nvjpegJpegStreamParseHeader(nvjpeg_handle_, sample->RawData<unsigned char>(),
+                                                  sample->Size(), jpeg_stream));
+            int is_supported = 0;  // 0 means yes
+            CUDA_CALL(nvjpegDecodeBatchedSupported(nvjpeg_handle_, jpeg_stream, &is_supported));
+            meta.can_decode = (is_supported == 0);
+            if (!meta.can_decode) {
+              promise.set(i, {false, nullptr});
+              parse_promise.set(i, {false, nullptr});
+              return;
+            }
+            unsigned int precision;
+            CUDA_CALL(nvjpegJpegStreamGetSamplePrecision(jpeg_stream, &precision));
+            meta.dyn_range_multiplier = 1.0f;
+            if (NeedDynamicRangeScaling(precision, DALI_UINT16)) {
+              meta.dyn_range_multiplier = DynamicRangeMultiplier(precision, DALI_UINT16);
+              meta.needs_processing = true;
+            }
+            meta.needs_processing = opts.dtype != DALI_UINT16;
+            if (!rois.empty() && rois[i].use_roi()) {
+              meta.needs_processing = true;
+            }
+            if (opts.use_orientation) {
+              auto &ori = meta.orientation = JpegParser().GetInfo(sample).orientation;
+              if (ori.rotate || ori.flip_x || ori.flip_y)
+                meta.needs_processing = true;
+            }
+            parse_promise.set(i, {true, nullptr});
+          } catch (...) {
+            meta.can_decode = false;
+            promise.set(i, {false, std::current_exception()});
+            parse_promise.set(i, {false, nullptr});
+          }
+        }, in[i]->Size());
+  }
+  ctx.tp->RunAll(false);
+  parse_promise.get_future().wait_all();
+
+  batch_sz_ = 0;
+  for (auto &meta : sample_meta_) {
+    if (meta.can_decode)
+      meta.idx_in_batch = batch_sz_++;
   }
 }
 
+void NvJpegLosslessDecoderInstance::RunDecode(DecodeContext ctx,
+                                              span<SampleView<GPUBackend>> out,
+                                              cspan<ImageSource *> in,
+                                              DecodeParams opts,
+                                              cspan<ROI> rois) {
+  if (batch_sz_ <= 0)
+    return;
+  int nsamples = in.size();
+  kernels::DynamicScratchpad s({}, ctx.stream);
```

Comment:
I think this breaks. The DynamicScratchpad is destroyed before any postprocessing on it invoked. Previously it was done in the same scope, now it is not.

Reply:
True, we've been here before.

Reply:
This kind of has me wondering if we can detect this somehow (without rewriting half of DALI kernels).